### ISSUE TITLE
False positive for function that takes a `Callable` as a parameter and returns a `Tuple` of `Union`

### ISSUE THREAD

**Issue by ZappD0S** [python/mypy#10281]
**Bug Report**

The following  code 
```python
from typing import (
    Callable,
    TypeVar,
    Union,
    Tuple,
)


T = TypeVar("T")


def test(pred: Callable[[T], bool]) -> Tuple[Union[T, int]]:
    return (1,)


def main(pred: Callable[[T], bool]) -> None:
    (val,) = test(pred)  # the error appears here


def f(x: int) -> bool:
    return x > 1


main(f)
```
triggers  the mypy error
> Argument 1 to "test" has incompatible type "Callable[[T], bool]"; expected "Callable[[Union[T, int]], bool]"

at the line that I marked with a comment. I believe this is a bug as this correction makes no sense to me (of couse, I could be wrong).
I'm aware that this code is meaningless, but it's the simplest sample I was able to produce that causes the error.
From the tests that I've done it seems to depend both on the type variable being used and on the function `test` returning a `Tuple[Union[...]]`. So if I replace `T` with a regular type of if I return either `Union[T, int]` or `Tuple[int]` instead of `Tuple[Union[T, int]]` the error goes away. It also disappears by not upacking the the result of `test(pred)`.

**To Reproduce**

Take the code above and check it with mypy. That's it.

**Expected Behavior**

No error triggered.

**Actual Behavior**

Mypy detects the error reported above.

**Your Environment**

- Mypy version used: 0.812
- Mypy command-line flags: --ignore-missing-imports --follow-imports=silent --show-column-numbers
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9.3
- Operating system and version: Linux x64
